pkg/distributors/contracts/MerkleOrchard.sol

```diff
@@ -0,0 +1,227 @@
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/MerkleProof.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IAsset.sol";
+
+import "./interfaces/IDistributor.sol";
+import "./interfaces/IDistributorCallback.sol";
+
+pragma solidity ^0.7.0;
+
+contract MerkleOrchard is IDistributor, Ownable {
+    using FixedPoint for uint256;
+    using SafeERC20 for IERC20;
+
+    // Recorded distributions
+    // rewardToken > rewarder > distribution > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => bytes32))) public trees;
+    // rewardToken > rewarder distribution > lp > root
+    mapping(IERC20 => mapping(address => mapping(uint256 => mapping(address => bool)))) public claimed;
+    // rewardToken > rewarder > balance
+    mapping(IERC20 => mapping(address => uint256)) public suppliedBalance;
+
+    IVault public immutable vault;
+
+    constructor(IVault _vault) {
+        vault = _vault;
+    }
+
+    struct Claim {
+        uint256 distribution;
+        uint256 balance;
+        address rewarder;
+        IERC20 rewardToken;
+        bytes32[] merkleProof;
+    }
+
+    function _processClaims(
+        address liquidityProvider,
+        address recipient,
+        Claim[] memory claims,
+        bool asInternalBalance
+    ) internal {
+        IVault.UserBalanceOpKind kind = asInternalBalance
+            ? IVault.UserBalanceOpKind.TRANSFER_INTERNAL
+            : IVault.UserBalanceOpKind.WITHDRAW_INTERNAL;
+        IVault.UserBalanceOp[] memory ops = new IVault.UserBalanceOp[](claims.length);
+
+        Claim memory claim;
+        for (uint256 i = 0; i < claims.length; i++) {
+            claim = claims[i];
+
+            require(
+                !isClaimed(claim.rewardToken, claim.rewarder, claim.distribution, liquidityProvider),
+                "cannot claim twice"
+            );
+            require(
+                verifyClaim(
+                    claim.rewardToken,
+                    claim.rewarder,
+                    liquidityProvider,
+                    claim.distribution,
+                    claim.balance,
+                    claim.merkleProof
+                ),
+                "Incorrect merkle proof"
+            );
+
+            require(
+                suppliedBalance[claim.rewardToken][claim.rewarder] >= claim.balance,
+                "rewarder hasn't provided sufficient rewardTokens for claim"
+            );
+
+            ops[i] = IVault.UserBalanceOp({
+                asset: IAsset(address(claim.rewardToken)),
+                amount: claim.balance,
+                sender: address(this),
+                recipient: payable(recipient),
+                kind: kind
+            });
```

Comment:
```suggestion
            ops[i] = IVault.UserBalanceOp({
                asset: IAsset(address(claim.rewardToken)),
                amount: claim.balance,
                sender: address(this),
                recipient: payable(recipient),
                kind: kind
            });
```

An issue I see with this is that for N claims we're performing N transfers. We should look to aggregate these.

Reply:
a difficulty in aggregating this is that we can't do in-memory mappings, so the aggregation logic is not particularly clean/comes at the expense of storage costs

Reply:
Had a stab at it here. Definitely isn't as pretty but I think it's worth the tradeoff: https://github.com/balancer-labs/balancer-v2-monorepo/pull/827